testing/src/ops_main_mock.py

```diff
@@ -0,0 +1,288 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import inspect
+import os
+import pathlib
+import sys
+from typing import TYPE_CHECKING, Any, Optional, Sequence, Type, cast
+
+import ops.charm
+import ops.framework
+import ops.model
+import ops.storage
+from ops import CharmBase
+
+# use logger from ops._main so that juju_log will be triggered
+try:
+    from ops._main import CHARM_STATE_FILE, _Dispatcher, _get_event_args  # type: ignore
+    from ops._main import logger as ops_logger  # type: ignore
+except ImportError:
```

Comment:
the imports here should be simplified now that we know what ops version we're being shipped with

Reply:
or, maybe not, since it's still an independent package that could be installed alongside older ops. nvm

Reply:
It's probably true that it could still theoretically be installed alongside older ops, but in that case why wouldn't you also just use an older ops-scenario version? I think part of the reason it's good in a single repo is so we can do atomic commits for things that use `_private` functions.

*However*, it might make sense to keep this PR limited to just copying code 1:1, and then subsequent PRs for updates to those files. Will make review saner I think, and mean we're not tempted to include other changes here.

Reply:
Yeah, I had to bump the minimum ops version in this PR so the compatibility code can be removed, but I would rather do it in another PR, since this one is already complicated to review.